README.md
# xlreg_rb

The Ruby client for the xlReg cluster registry.

**xlreg_rb**
makes it easy to create new small clusters some or all of whose members are
written in the programming language Ruby.
xlreg_rb is expected to be used primarily for testing, where
the ability to create and launch new, unique, distinct clusters in seconds
will be a major advantage.

In such a cluster, members use the **xlReg** public server to register and
exchange member IP addresses and other related information.
When a member finishes booting, it has
learned the cluster ID and the
IP addresses of its peers, their NodeIDs, and the public
RSA keys used to set up
encrypted communications
between members of the cluster and between cluster members and clients.

## The xlReg Service

<img src="img/xl-registration.jpg" alt="xl-registration" style="float:left" title="members registering with xlReg">

[xlReg](http://jddixon.github.io/xlReg_go)
is a service which provides its clients with unique nodeIDs and allows them
to register and join new clusters and collect configuration information
about existing clusters.

An **xlReg cluster** is a number of machines
which cooperate as a set of intercommunicating servers.  Each
such server has two RSA keys.  One, the **sig** key, is used for creating
digital signatures.  The other, the **comms** key, is used for encrypting
communications.  Most often the comms key is used for exchanging keys for
block ciphers. That is, the comms key is used only for agreeing on the
much faster block cipher key used during the rest of the communications
session.

Each xlReg cluster member has at least one IP address used for
intra-cluster communications, for communications between members.  In some
clusters each member will also have a second IP address used for
communications with cluster clients.

The xlReg server itself, its clients, and so the cluster members all are
or behave like
[XLattice nodes](http://jddixon.github.io/xlNode_go).
That is, they are self-contained fault-tolerant entities with some
persistent store.  These nodes may be located anywhere on the global
Internet, or on private networks with access to the network in which
an xlReg server resides.`

Like any XLattice node, the xlReg client can either be configured with some
specific keys at start-up or it will generate a new pair of keys and save
them to persistent store.

`## Registration

<img src="img/simple-cluster.jpg" alt="simple-cluster" style="float:right" title="small cluster, no clients">

A booting xlReg client establishes an encrypted connection to an xlReg server
(by default this is xlreg.xlattice.net).
If this is the client's first conversation with the xlReg server, the client
provides the server with configuration information including its RSA public
keys.  (The client does **not** register its secret RSA **private** keys.)
The configuration information is signed with the applicant's private
*sig" key.  The server uses the
client's RSA public key, provided as part of the configuration information, to
verify the client's digital signature on the configuration data.  This proves
that the configuration information is intact and that the applicant has the
private key corresponding to the public key.  In other words, the server has
in a certain sense confirmed the applicant's identity.  Having done this,
the xlReg server issues the client with a unique node ID, a 20-byte value,
completing client registration.

Once an xlReg registry client has identified itself to the registry it can

1. register new clusters or
2. join existing clusters or
3. get information on existing clusters.

The third option is used, for example, by nodes intending to become clients
of an existing cluster.

## Acquiring Configuration Data

After a new member has joined a cluster the server provides it
with cluster configuration data.  This includes

* the cluster name,
* the cluster ID, and
* the cluster size, the  maximum number of cluster members

It also includes for each current cluster member the member's

* nodeID,
* IP address(es),
* **sig** RSA public key, and
* **comms** RSA public key

<img src="img/cluster-with-clients.jpg" alt="cluster-with-clients" style="float:left" title="cluster with clients">

The new cluster member can then use this information to communicate with
other members.  Cluster members need no further assistance from the
xlReg server.

Cluster members will communicate with one another and clients
using protocols agreed-upon among themselves, but the xlReg registry
has no knowledge of such protocols.

**xlreg_rb** is currently in development.  This is not
necessarily a description of running code; it is where we are going.

xlReg is an [XLattice](http://jddixon.github.io/xlattice_go/) project.  More
detailed (but somewhat dated) conceptual information on XLattice
is available at the [2003-2006 Java XLattice website](http://www.xlattice.org).

xlReg servers and clients communicate using the
[Gooble Protocol Buffers](http://code.google.com/p/protobuf/)-based
[xlReg protocol](http://jddixon.github.io/xlReg_go/xlReg_protocol.html).
An xlReg server has been in test for several months.

## Public Development Server

There is currently an xlReg server running on xlreg.xlattice.org:56789,
with the public keys shown in the registry credentials below.

    regCred {
        Name: xlReg
        ID: f2384769af97880208e76d012d8cf5b1b8be0b7b
        CommsPubKey: ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDH6nJcf4cAbb1RWVvNS6YHjatZJQCerhmISdyhBDmNJAUuBpr+LcrOe0d+kxEPQB+IbNKfvj7ROd8Aet74s5sJ0AcZyFXLFPvt6eugLCzrXmgDo60qoUtZgxaxisQvsoAZB2FR+nqJAG8E84o3rGD1bNM/2WgjvuWoLpfy+RE8hCUFku7yo9JdOunhc4rpCkBeX9M7g8oWPanHDWMrSwsE8CBf/4pXlEiNNd3YHkwQseUrkCvpLCDuYYV8xjzAYexoBWrVGwpeTO/QtwdebwPxa7ArEQfPwgA5hn6S9iy9+EK6zay7OKqKioaOoYDlJ+Z1rAIi6dBJXdRVHGsMc/kl

        SigPubKey: ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQCk84/cQ/YnxOLElukZpeQ9QT9FdgcYujmIUGOaFcWtxUJJoSXgBterjgJ+fE5vvNN/kdX97Ps9Mz8PHTAUyClUn5DMxFYApDp/UwnBk0MNqaCnqt3p9ymflIZiex1+/pSWcSpqZYW9hIUQddEGWG6a4dWyD2lhoD1II5Iuj0xMgN167fgsA7otgd/wTitp2jrJE+0cZ4eQShpkuMd9BvNJmkMoyF5PzBnMcRxZ9RvM4zEOe23GRSGGAuPUvTaVEeYY3TVFgF4YfXiYOrPQynS6B7+fnWXu2XDQrwP3mZxYAYJqL5nZvQypCilc8dDxFunf5B+GxdAwqyQbTNwM3ksH

        EndPoints {
             TcpEndPoint: 54.186.197.123:56789
        }
        Version: 1.9.20
    }

The IP address in the reg credentials above is of course that of xlreg.xlattice.org.

## Server Implementation

The xlReg server and its supporting libraries are written in
[the Go programming language](http://golang.org), a high-performance
systems programming language incorporating certain features of
[Tony Hoare](http://en.wikipedia.org/wiki/Tony_Hoare)'s
revolutionary **CSP**,
[Communicating Sequential Processes](http://www.usingcsp.com).

Google's Go language is a stable product.  It has been available for
several years and the compiler is widely acknowledged to be very fast
and to generate fast, compact code.  The Go language itself --
the compiler and its libraries -- is free and open source.

## Server Source Code

Software developers can currently obtain the xlReg server source code at
[xlReg_go's Github project page](https://github.com/jddixon/xlReg_go/)
so that if you are impatient to begin using xlReg you can build and operate
your own server.

## Client Source Code

This is best obtained through
[this client's Github project page](https://github.com/jddixon/xlreg_rb/).

## Project Status

The xlReg server is currently a good beta: it is reasonably reliable and
there are no known serious bugs.

**This client is itself pre-alpha:** some minimal code exists, but nothing
beyond reading the registry credentials (`regCred.dat`) and going through
the initial hello/reply session with the server, xlReg.
## On-line Documentation

More information on the **xlreg_rb** project can be found
[here](https://jddixon.github.io/xlreg_rb)


CHANGES
~/dev/ruby/xlreg_rb/CHANGES

v0.0.11
    2016-11-17
        * updae web pages for xlReg 1.9.20                          SLOC 1036
v0.0.10
    2016-09-24
        * standardize licensing                                     SLOC 1034
v0.0.9
    2016-07-15
        * add ghpDoc/{css,img}                                      SLOC 1008
v0.0.8
    2016-06-13
        * add .dvcz, bkp2U, other bash scripts                      SLOC  498
v0.0.7
    2016-03-02
        * update LICENSE                                            SLOC  242 
v0.0.6
    2015-12-22
        * fix bad link
v0.0.5
    2015-12-07
        * add version.rb; minor edits to *.md                       SLOC  242
v0.0.4
    2015-10-15
        * update README/index.md for xlReg 1.9.10                   SLOC  240
v0.0.3
    2015-08-05
        * update README/index.md for xlReg 1.9.9                    SLOC  240
v0.0.2
    2015-06-05
        * updated docs for regCred.dat 1.9.6                        SLOC  240
v0.0.1
    2015-04-19
        * clean up initial version                                  SLOC  240









errors_test.rb

require 'test/unit'
require './xlreg_errors.rb'

class TestErrors < Test::Unit::TestCase

  def testBasic
    # DEBUG
    sleep 0.010
    $stderr.print("START TEST RUN\n")
    sleep 0.010
    # END DEBUG
    
    begin
      # DEBUG
      err = BadAttrsLine.new('some message')  # not seen
      sleep 0.010
      # END DEBUG
      raise err, 'additional comment'         # seen
      flunk ('raised BadAttrsLine but not caught')

    rescue BadAttrsLine
      # DEBUG
      $stdout.print("BadAttrsLine has been caught\n")   # DEBUG
      sleep 0.010
      # END DEBUG
      msg = $!.message
      #             expected,                           actual
      assert_equal('bad attrs line: additional comment', msg)
    
    rescue XLRegException 
      msg = $!.message
      #             expected,                           actual
      assert_equal('bad attrs line: additional comment', msg)
    
    rescue  StandardError
      msg = $!.message
      #             expected,                           actual
      assert_equal('bad attrs line: additional comment', msg)
    end
  end
end


p.pb.rb
### Generated by rprotoc. DO NOT EDIT!
### <proto file: p.proto>
# package reg;
# 
# // All messages are encrypted using the server-provided
# // session key and PKCS7-padded.
# 
# message XLRegMsg {
#     enum Tag {
#         RegCredRequest  = 1;   
#         RegCredReply    = 2;    // version, token (other registry credentials)
# 
#         Client      	= 3;    // token OR clientID, digSig
#         ClientOK    	= 4;    // clientID, attrs
# 
#         // size means cluster size, constrained to 1 < n <= 64
#         // epCount is number of menber endPoints, which must be 1 or 2
#         Create      	= 5;    // clusterName, size, epCount
#         CreateReply 	= 6;    // clusterID, size, epCount
# 
#         // Response to Join is clusterID + size + epCount -OR- error
#         Join        	= 7;    // clusterName or clusterID
#         JoinReply   	= 8;    // clusterID, attrs, size, epCount
# 
#         GetCluster  	= 9;    // clusterID, which = bit vector; -1 = all
#         // Response to Get is a list of known members -OR- error
#         ClusterMembers 	= 10;   // clusterID, which, tokens
# 
#         Bye         	= 11;   // from client
#         Ack         	= 12;   // from server; followed by close
#         
#         Error       	= 13;   // errDesc; from server, followed by close
#         
#     }
#     // the Token describes a member
#     message Token {
#         optional string Name        = 1;
# 	    optional uint64 Attrs       = 2;    // bit field
# 	    optional bytes  ID          = 3;    // 20 or 32 byte nodeID
#         optional bytes  CommsKey    = 4;
#         optional bytes  SigKey      = 5;
#         // by convention, MyEnds[0] for inter-cluster comms, 
#         // MyEnds[1] for cluster-client comms
#         // there must be epCount endPoints present
#         repeated string MyEnds      = 6;    // overlay, endPoint
# 	    optional bytes  DigSig      = 7;    // over fields present, in order
# 	}
#     optional Tag    Op              = 1;
#     optional bytes  AesIV           = 2;
#     optional bytes  AesKey          = 3;
#     optional bytes  Salt1           = 4;    // error if len < 8
#     optional bytes  Salt2           = 5;    // error if len < 8
#     optional uint32 Version         = 6;    // little-endian, so stored D.C.B.A
#     
#     optional string ClientName      = 8;
#     optional bytes  ClientID        = 9;
#     optional uint64 ClientAttrs     = 10;
#     optional Token  ClientSpecs     = 11;
# 
#     optional bytes  ClusterID       = 13;       
#     optional string ClusterName     = 14; 
#     optional uint32 ClusterSize     = 15;
#     optional uint64 ClusterAttrs    = 16; 
#     optional uint32 EndPointCount   = 17;   // each member must have
# 
#     optional uint64 Which           = 20;   // bitset, members requested/sent 
#     repeated Token  Tokens          = 21;   // specs for members
#     optional bytes  DigSig          = 22;
# 
#     optional string ErrDesc         = 23;
# 
# }
# 

require 'protobuf/message/message'
require 'protobuf/message/enum'
require 'protobuf/message/service'
require 'protobuf/message/extend'

module Reg
  class XLRegMsg < ::Protobuf::Message
    defined_in __FILE__
    class Tag < ::Protobuf::Enum
      defined_in __FILE__
      RegCredRequest = value(:RegCredRequest, 1)
      RegCredReply = value(:RegCredReply, 2)
      Client = value(:Client, 3)
      ClientOK = value(:ClientOK, 4)
      Create = value(:Create, 5)
      CreateReply = value(:CreateReply, 6)
      Join = value(:Join, 7)
      JoinReply = value(:JoinReply, 8)
      GetCluster = value(:GetCluster, 9)
      ClusterMembers = value(:ClusterMembers, 10)
      Bye = value(:Bye, 11)
      Ack = value(:Ack, 12)
      Error = value(:Error, 13)
    end
    class Token < ::Protobuf::Message
      defined_in __FILE__
      optional :string, :Name, 1
      optional :uint64, :Attrs, 2
      optional :bytes, :ID, 3
      optional :bytes, :CommsKey, 4
      optional :bytes, :SigKey, 5
      repeated :string, :MyEnds, 6
      optional :bytes, :DigSig, 7
    end
    optional :Tag, :Op, 1
    optional :bytes, :AesIV, 2
    optional :bytes, :AesKey, 3
    optional :bytes, :Salt1, 4
    optional :bytes, :Salt2, 5
    optional :uint32, :Version, 6
    optional :string, :ClientName, 8
    optional :bytes, :ClientID, 9
    optional :uint64, :ClientAttrs, 10
    optional :Token, :ClientSpecs, 11
    optional :bytes, :ClusterID, 13
    optional :string, :ClusterName, 14
    optional :uint32, :ClusterSize, 15
    optional :uint64, :ClusterAttrs, 16
    optional :uint32, :EndPointCount, 17
    optional :uint64, :Which, 20
    repeated :Token, :Tokens, 21
    optional :bytes, :DigSig, 22
    optional :string, :ErrDesc, 23
  end
end

decimal_version.rb
# xlreg_rb/decimal_version.rb

require 'bindata'

# See bindata wiki.  Note particularly the array: f field in the example,
# which is :uint32.  Also note how PascalStringReader is implemented,
# with the length returned by :read_length

class DecimalVersion < BinData::Record
  endian :big
  uint32 :val

  def initialize(aIn, bIn, cIn, dIn)
    raise if aIn.nil?
    raise unless aIn.is_a(Numeric)
    raise unless (0 <= aIn) && (aIn < 256)
    a = aIn

    if bIn.nil? 
      a = 0
    else  
      raise unless bIn.is_a(Numeric)
      raise unless (0 <= bIn) && (bIn < 256)
      b = bIn
    end

    if cIn.nil?
      c = nil
    else
      raise unless cIn.is_a(Numeric)
      raise unless (0 <= cIn) && (cIn < 256)
      b = bIn
      c = cIn
    end

    if dIn.nil?
      d = 0
    else
      raise unless dIn.is_a(Numeric)
      raise unless (0 <= dIn) && (dIn < 256)
      d = dIn
    end

    @val = a | (b << 8) | (c << 16) | (d << 24)     
  end

  def getA()
    return @val & 0xff
  end

  def getB()
    return (@val >> 8) & 0xff
  end

  def getC()
    return (@val >> 16) & 0xff
  end

  def getD()
    return (@val >> 24) & 0xff
  end

end


xlreg_errors.rb
# xlreg_rb/xlreg_errors.rb

module XLRegErrors

  BAD_ATTRS_LINE = -1
  BAD_VERSION = -2
  CANT_FIND_CLUSTER_BY_ID = -3
  CANT_FIND_CLUSTER_BY_NAME = -4
  CLUSTER_MEMBER_ID_IN_USE = -5
  CLUSTER_MEMBERS_MUST_HAVE_END_POINT = -6
  CLUSTER_MEMBER_NAME_IN_USE = -7
  CLUSTER_MUST_HAVE_MEMBER = -8
  CLUSTER_MUST_HAVE_TWO = -9
  ID_ALREADY_IN_USE = -10
  ILL_FORMED_CLUSTER = -11
  ILL_FORMED_CLUSTER_MEMBER = -12
  ILL_FORMED_REG_CRED = -13
  MEMBER_MUST_HAVE_END_POINT = -14
  MISSING_CLOSING_BRACE = -15
  MISSING_CLUSTER_NAME_OR_ID = -16
  MISSING_END_POINTS_SECTION = -17
  MISSING_MEMBERS_LIST = -18
  MISSING_NODE = -19
  MISSING_PRIVATE_KEY = -20
  MISSING_REG_NODE_LINE = -21
  MISSING_SERVER_INFO = -22
  NAME_ALREADY_IN_USE = -23
  NIL_CLUSTER = -24
  NIL_CLUSTER_MEMBER = -25
  NIL_NODE = -26
  NIL_PRIVATE_KEY = -27
  NIL_REGISTRY = -28
  NIL_REG_NODE = -29
  NIL_TOKEN = -30
  RCVD_INVALID_MSG_FOR_STATE = -31
  TAG_OUT_OF_RANGE = -32
  UNEXPECTED_MSG_TYPE = -33
  WRONG_NUMBER_OF_BYTES_IN_ATTRS = -34


  XLREG_ERRORS = [ \
    'badly formed attrs line',
    'badly formated VERSION',
    'cannot find cluster with this ID',
    'cannot find cluster with this name',
    'cluster member id already in use',
    'cluster members must have at least one endPoint',
    'cluster member name already in use',
    'cluster must have a member!',
    'cluster must have at least two members',
    'ID already in use',
    'ill-formed cluster serialization',
    'ill-formed cluster member serialization',
    'ill-formed regCred serialization',
    'member must have at least one endPoint',
    'missing closing brace',
    'missing cluster name or ID',
    'missing endPoints section',
    'missing members list',
    'missing node parameter',
    'missing private key line',
    'missing regNode line',
    'missing server info',
    'name already in use',
    'nil cluster argument',
    'nil cluster member argument',
    'nil node argument',
    'nil private key argument',
    'nil registry argument',
    'nil RegNode argument',
    'nil XLRegMsg_Token argument',
    'invalid msg type for current state',
    'message tag out of range',
    'unexpected message type',
    'wrong number of bytes in attrs',
  ]
end
